apps/asap-server/test/handlers/calendar/subscribe-handler.test.ts

```diff
@@ -123,44 +127,74 @@ describe('Calendar Webhook', () => {
       expect(subscribe).not.toHaveBeenCalled();
     });
 
-    test('Should unsubscribe and remove the resourceId then resubscribe if the calendar ID changed', async () => {
+    describe('Calendar ID changed', () => {
       const resourceId = 'some-resource-id';
       const expiration = 123456;
-      const calendarUpdateEvent = getCalendarUpdateEvent();
-      subscribe.mockResolvedValueOnce({ resourceId, expiration });
+      test('Should unsubscribe and remove the resourceId then resubscribe', async () => {
+        const calendarUpdateEvent = getCalendarUpdateEvent();
+        subscribe.mockResolvedValueOnce({ resourceId, expiration });
+
+        const res = await handler(
+          getEvent('CalendarUpdated', calendarUpdateEvent),
+        );
+
+        expect(res).toBe('OK');
+        expect(unsubscribe).toHaveBeenCalledWith(
+          calendarUpdateEvent.payload.dataOld!.resourceId!.iv,
+          calendarUpdateEvent.payload.id,
+        );
+        expect(calendarControllerMock.update).toHaveBeenCalledWith(
+          calendarUpdateEvent.payload.id,
+          {
+            resourceId: null,
+          },
+        );
+        expect(subscribe).toHaveBeenCalled();
+      });
 
-      const res = await handler(
-        getEvent('CalendarUpdated', calendarUpdateEvent),
-      );
+      test('Should skip subscription and unsubscribing and return 200 when the version is old', async () => {
```

Comment:
are we sure you want to skip unsubscribing if the version is old? technically this means that if we need to unsubscribe but a new version has been saved we will never be able to unsubscribe. Or maybe we dont need to?

Reply:
The unsubscribe is in 2 parts:

1. `await unsubscribe(payload.dataOld.resourceId?.iv, payload.id);`
2. ```await calendarController.update(payload.id, {
            resourceId: null,
          });```

The first unsubscribes from Google, I think we could do this.
The second sets the resourceid to null in squidex. Could we get in a senario where a later event has been applied, and the resourceid updated to a newer one?

Reply:
i guess in the current setup we cannot - unless of course we subscribe from one valid calendar to another but that is a whole different scenario of course because then there will be no race condition and no retry attempt to unsubscribe and subscribe to an invalid (fake) calendar (which is what the admins do now)

If they 1) Change the calendar to a non-existent calendarId and then 2) Set it again to the same calendarId as at the beginning  

Then if 2) comes in first the system will detect no difference in the calendarIds (there is a comparison in the code right?) so it will not try to unsubscribe, in fact it will exit immediately

Once 1) comes in, in the old code it would then unsubscribe the calendar and try to subscribe to the new one which would fail resulting in empty subscription (can we test that? maybe there is no point) but with your improvement it should reject the update all together, right?


Question is - can we test it? Without actually running this against squidex it's a silly test really but we might be able to invoke the handler in one test twice with different input to replicate the out-of-order event issue and write some expectations that would at least describe what we are trying to do, what do you reckon?

Reply:
```2) comes in first the system will detect no difference in the calendarIds (there is a comparison in the code right?)```

There is a comparision in the code, but in step 2 it will compare the valid calendar Id with the invalid one. Leading to it being unsubscribed.

https://github.com/yldio/asap-hub/pull/1153/commits/aab9a64ab168416f4fe7d7b6416b0a221aa65f17